apollo-federation/src/source_aware/federated_query_graph/path_tree.rs

```diff
@@ -27,8 +32,255 @@ pub(crate) struct Child {
+                by_unique_trigger: IndexMap::new(),
+            });
+
+            match for_edge
+                .by_unique_trigger
+                .entry(edge.operation_element.clone())
+            {
+                indexmap::map::Entry::Occupied(existing) => {
+                    let existing = existing.into_mut();
+                    merge_and_remap_condition_resolution_ids(
+                        &mut existing.self_condition_resolutions_for_edge,
+                        &edge.self_condition_resolutions_for_edge,
+                        remapped_condition_ids,
+                    );
+                    existing
+                        .sub_paths_and_selections
+                        .push((graph_path_iter, selection));
+                }
+                indexmap::map::Entry::Vacant(vacant) => {
+                    vacant.insert(PathTreeChildInputs {
+                        self_condition_resolutions_for_edge: edge
+                            .self_condition_resolutions_for_edge
+                            .clone(),
+                        sub_paths_and_selections: vec![(graph_path_iter, selection)],
+                    });
+                }
+            }
+        }
+
+        let mut childs = vec![];
+        for (edge, by_unique_edge) in merged {
+            for (operation_element, child) in by_unique_edge.by_unique_trigger {
+                childs.push(Arc::new(Child {
+                    key: ChildKey {
+                        operation_element,
+                        edge,
+                    },
+                    self_condition_resolutions_for_edge: child.self_condition_resolutions_for_edge,
+                    tree: Self::from_paths_inner(
+                        graph.clone(),
+                        by_unique_edge.target_node,
+                        child.sub_paths_and_selections.into_iter(),
+                        remapped_condition_ids,
+                    )?
+                    .into(),
+                }))
+            }
+        }
+
+        Ok(Self {
+            graph,
+            node,
+            condition_resolutions_at_node,
+            source_entering_condition_resolutions_at_node,
+            childs,
+        })
+    }
+
+    /// Create a path tree by merging the given graph paths.
+    /// XXX(@goto-bus-stop): Do we need to keep the FederatedGraphPaths around independently of the
+    /// path tree? Or can we just take ownership in this function?
+    pub fn from_paths<'a>(
+        graph: Arc<FederatedQueryGraph>,
+        node: NodeIndex,
+        paths: impl Iterator<Item = (&'a FederatedGraphPath, Option<Arc<SelectionSet>>)>,
+    ) -> Result<Self, FederationError> {
+        Self::from_paths_inner(
+            graph,
+            node,
+            paths.map(|(path, selection_set)| (path.edges(), selection_set)),
+            &mut Default::default(),
+        )
+    }
+
+    // Path trees should not normally be cloned. This manual implementation is private
```

Comment:
I'm not sure if this is true. I just didn't want to add the public API without being sure.